# Основное задание: Изучение библиотечных функций

### Требования: keras.optimizers, torch.optim: SGD, и модификации SGD (Nesterov, Momentum, AdaGrad, RMSProp, Adam). Изучите параметры вызываемых библиотечных функций.

# Изучение библиотеки _torch_

##### Выбираем как _target_ alcohol и обучаем модель на датасете

### ***SGD***:

In [19]:
from libs.chooseLib import choose_model
import numpy as np

if __name__ == '__main__':
    sample_raw = np.asarray(
            [[7.4, 0.70, 0.00, 1.9, 0.076, 11, 34, 0.9978, 3.51, 0.56, 5]]
        )

    model, scaler = choose_model('alcohol',"torch", "SGD")
    sample_scaled = scaler.transform(sample_raw)
    print(model(sample_scaled))

[DataLoader] shape=(1599, 12)
MSE: 0.35289, MAE: 0.45928, R²: 0.71654
9.618028


### ***Momentum***:

In [20]:
    model, _ = choose_model('alcohol',"torch", "Momentum")
    print(model(sample_scaled))

[DataLoader] shape=(1599, 12)
MSE: 0.34626, MAE: 0.45549, R²: 0.72186
9.651606


### ***Nesterov***:

In [21]:
    model, _ = choose_model('alcohol',"torch", "Nesterov")
    print(model(sample_scaled))

[DataLoader] shape=(1599, 12)
MSE: 0.34513, MAE: 0.45151, R²: 0.72277
9.624086


### ***RMSProp***:

In [22]:
    model, _ = choose_model('alcohol',"torch", "RMSProp")
    print(model(sample_scaled))

[DataLoader] shape=(1599, 12)
MSE: 0.37037, MAE: 0.45557, R²: 0.70250
9.425782


### ***Adam***:

In [23]:
    model, _ = choose_model('alcohol',"torch", "Adam", lr=0.5)
    print(model(sample_scaled))

[DataLoader] shape=(1599, 12)
MSE: 0.34831, MAE: 0.45912, R²: 0.72021
9.675847


### ***AdaGrad***:

In [24]:
    model, _ = choose_model('alcohol',"torch", "AdaGrad", lr=0.9999991)
    print(model(sample_scaled))

[DataLoader] shape=(1599, 12)
MSE: 0.42791, MAE: 0.48814, R²: 0.65628
9.327294


---
_Real Value: 9.4_

| Optimizer    |     MSE      |     MAE      |     R²      | Predict Value (s) |
|--------------|--------------|--------------|-------------|-------------------|
| SGD          |   0.35582    |   0.46175    |   0.71418   | 9.615764          |
| Momentum     |   0.34623    |   0.45548    |   0.72189   | 9.652524          |
| Nesterov     |   0.34510    |   0.45149    |   0.72279   | 9.624092          |
| RMSProp      |   0.36799    |   0.45324    |   0.70441   | 9.463510          |
| Adam         |   0.34945    |   0.46032    |   0.71930   | 9.677068          |
| AdaGrad      |   0.47579    |   0.51806    |   0.61781   | 9.254324          |

---

# Изучаем библиотеку _keras_

### ***SGD***:

In [25]:
    model, _ = choose_model('alcohol',"tensorflow", "SGD")
    print(model(sample_scaled))

[DataLoader] shape=(1599, 12)


2025-05-30 15:46:21.963569: E external/local_xla/xla/stream_executor/cuda/cuda_platform.cc:51] failed call to cuInit: INTERNAL: CUDA error: Failed call to cuInit: UNKNOWN ERROR (303)
/home/letooow/PycharmProjects/PythonProject/.venv/lib/python3.12/site-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 0, Loss: 0.427412063
SGD: MSE = 0.3468, MAE = 0.4530, R² = 0.7214
9.589417


### ***Momentum***:

In [26]:
    model, _ = choose_model('alcohol',"tensorflow", "Momentum")
    print(model(sample_scaled))

[DataLoader] shape=(1599, 12)


/home/letooow/PycharmProjects/PythonProject/.venv/lib/python3.12/site-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 0, Loss: 23.6407337
Momentum: MSE = 0.6211, MAE = 0.5900, R² = 0.5011
9.90852


### ***Nesterov***:

In [27]:
    model, _ = choose_model('alcohol',"tensorflow", "Nesterov")
    print(model(sample_scaled))

[DataLoader] shape=(1599, 12)


/home/letooow/PycharmProjects/PythonProject/.venv/lib/python3.12/site-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 0, Loss: 0.54085505
Nesterov: MSE = 0.4023, MAE = 0.4855, R² = 0.6769
9.602156


### ***AdaGrad***:

In [31]:
    model, _ = choose_model('alcohol',"tensorflow", "AdaGrad", lr=0.9999991)
    print(model(sample_scaled))

[DataLoader] shape=(1599, 12)


/home/letooow/PycharmProjects/PythonProject/.venv/lib/python3.12/site-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 0, Loss: 18.9184189
AdaGrad: MSE = 0.3489, MAE = 0.4539, R² = 0.7198
9.646044


### ***RMSProp***:

In [29]:
    model, _ = choose_model('alcohol',"tensorflow", "RMSProp")
    print(model(sample_scaled))

[DataLoader] shape=(1599, 12)


/home/letooow/PycharmProjects/PythonProject/.venv/lib/python3.12/site-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 0, Loss: 62.4272346
RMSProp: MSE = 0.3856, MAE = 0.4913, R² = 0.6903
9.903184


### ***Adam***:

In [30]:
    model, _ = choose_model('alcohol',"tensorflow", "Adam")
    print(model(sample_scaled))

[DataLoader] shape=(1599, 12)


/home/letooow/PycharmProjects/PythonProject/.venv/lib/python3.12/site-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 0, Loss: 71.9465103
Adam: MSE = 0.3992, MAE = 0.4837, R² = 0.6793
9.348383


### Результат:

| Optimizer |   MSE   |   MAE   |   R²   | Predict Value |
|-----------|---------|---------|--------|---------------|
| SGD       | 0.3468  | 0.4530  | 0.7214 | 9.589417      |
| Momentum  | 0.6211  | 0.5900  | 0.5011 | 9.908520      |
| Nesterov  | 0.4023  | 0.4855  | 0.6769 | 9.602156      |
| AdaGrad   | 0.3489  | 0.4539  | 0.7198 | 9.646044      |
| RMSProp   | 0.3856  | 0.4913  | 0.6903 | 9.903184      |
| Adam      | 0.3992  | 0.4837  | 0.6793 | 9.348383      |



# Вывод



| Optimizer |  MSE (TF) | MSE (Torch) | Better MSE | R² (TF) | R² (Torch) | Better R² | MAE (TF) | MAE (Torch) | Better MAE | TF Predict | Torch Predict |
|-----------|-----------|-------------|------------|---------|------------|-----------|----------|-------------|------------|-------------|----------------|
| SGD       | 0.3468    | 0.35582     | TF         | 0.7214  | 0.71418    | TF        | 0.4530   | 0.46175     | TF         | 9.589417    | 9.615764       |
| Momentum  | 0.6211    | 0.34623     | Torch      | 0.5011  | 0.72189    | Torch     | 0.5900   | 0.45548     | Torch      | 9.908520    | 9.652524       |
| Nesterov  | 0.4023    | 0.34510     | Torch      | 0.6769  | 0.72279    | Torch     | 0.4855   | 0.45149     | Torch      | 9.602156    | 9.624092       |
| AdaGrad   | 0.3489    | 0.47579     | TF         | 0.7198  | 0.61781    | TF        | 0.4539   | 0.51806     | TF         | 9.646044    | 9.254324       |
| RMSProp   | 0.3856    | 0.36799     | Torch      | 0.6903  | 0.70441    | Torch     | 0.4913   | 0.45324     | Torch      | 9.903184    | 9.463510       |
| Adam      | 0.3992    | 0.34945     | Torch      | 0.6793  | 0.71930    | Torch     | 0.4837   | 0.46032     | Torch      | 9.348383    | 9.677068       |



| Категория                | **Keras (`tf.keras`)**                          | **PyTorch**                                     |
| ------------------------ | ----------------------------------------------- | ----------------------------------------------- |
| **Уровень абстракции**   | Высокоуровневый (ближе к "конструктору")        | Среднеуровневый (почти "ручной")                |
| **Простота кода**        | Максимально простой (`model = Sequential(...)`) | Чуть сложнее, нужно писать классы               |
| **Гибкость**             | Ограничена архитектурой Keras                   | Полная: кастомные forward, оптимизаторы, лоссы  |
| **Обучение**             | Автоматическое через `.fit()`                   | Явное через `for epoch in range(...)`           |
| **Отладка**              | Умеренная (сложно лезть внутрь `.fit()`)        | Отличная (можно пошагово дебажить `forward()`)  |
| **Визуализация**         | Встроенная: `TensorBoard`, `model.summary()`    | Через `TensorBoard`, `wandb`, `matplotlib`      |
| **Готовые модели**       | Очень много (`tf.keras.applications`)           | Много (`torchvision.models`, `torch.hub`)       |
| **Совместимость**        | Плотно связан с TensorFlow                      | Независим, но интегрируется с ONNX              |
